### PR TITLE
lindsay - edges - mediaranker

### PR DESCRIPTION
**Pull request by elle-terch**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | In Work, I created a class method to tally total votes by work and then rank it from most votes to least. I used a query to take the votes table, join it with the works table and then group it by work.id. From there I used order to sort it from most votes to least (and had to troubleshoot some weird SQL stuff). 
Describe how you approached testing that model method. What edge cases did you come up with? | I wrote the methods first. Once I had them working locally, I went through each method and created tests for each of them. I did not come up with edge cases. 
What are `session` and `flash`? What is the difference between them? | Session and flash are special rails hashes. Flash displays a message to a user whereas session assigns a session id each time a user logs in (it persists until the user logs out). 
Describe a controller filter you wrote. | I didn't get around to writing controller filters.
What was one thing that you gained more clarity on through this assignment? | In general, I gained more understanding of how all of the rails pieces fit together. 
What is the Heroku URL of your deployed application? | https://terch-mediaranker.herokuapp.com/
Do you have any recommendations on how we could improve this project for the next cohort? | Maybe some more guidance on how much testing is expected for models. I did appreciate the open-ended nature of the project because it forced me to figure out what things to prioritize.

### CODE REVIEW

File: app/controllers/pages_controller.rb [hunk 1]
```diff
@@ -0,0 +1,12 @@
+class PagesController < ApplicationController
+
+  def index
+
+    @top_books = Work.top_ten("book")
+    @top_movies = Work.top_ten("movie")
+    @top_albums = Work.top_ten("album")
+
+    @top_media = Work.top_media
```
Comment (droberts-sea) [inline app/controllers/pages_controller.rb:9]: Your custom model methods make this controller action very clean - good work!

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class User < ApplicationRecord
+  has_many :votes
+
+  def media_votes
+    combined_works = []
+
+    self.votes.each do |vote|
```
Comment (droberts-sea) [inline app/models/user.rb:7]: Why not use a `has_many :through` relationship for this?

File: app/models/vote.rb [hunk 3]
```diff
@@ -0,0 +1,5 @@
+class Vote < ApplicationRecord
+  belongs_to :user
+  belongs_to :work
+
```
Comment (droberts-sea) [inline app/models/vote.rb:4]: If you wanted to make it so that each user can only vote for a given work once, you could do it with a validation here.

File: app/models/work.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    category = category.downcase
+    all_media = self.where(category: category)
+    votes_by_work = all_media.total_votes
+  end
+
+
+  def self.top_ten(category)
+    top_ten = self.ranked_media(category)
+    top_ten.first(10)
+  end
```
Comment (droberts-sea) [inline app/models/work.rb:21]: I love the way the methods in this model build on each other little by little - this is a great example of functional decomposition.

File: app/views/works/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+      <% end %>
+
+    </tbody>
+  </table>
+</section>
+
+<section class = "media-table">
+  <h4>Books</h4>
+  <table>
+    <thead>
```
Comment (droberts-sea) [inline app/views/works/index.html.erb:35]: You have the same code to show a list of works repeated 3 times. Could you use a view partial or a loop to DRY this up?

File: app/views/works/edit.html.erb [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<h2>Edit this Work</h2>
+
+
+<%= form_with model: @work do |f| %>
```
Comment (droberts-sea) [inline app/views/works/edit.html.erb:4]: You have the same form code here and in `new.html.erb`. You should be using a partial to DRY this up.

File: config/routes.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  post '/logout', to: 'users#logout', as: 'logout'
+
+  resources :works do
+    member do
+      post 'upvote'
+    end
+  end
+
+  resources :users
+
```
Comment (droberts-sea) [inline config/routes.rb:19]: Do you need all 7 RESTful routes for `user`?

File: test/models/work_test.rb [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+  describe 'top_ten' do
+
+    it 'returns only 10 entries of each category' do
+
+      x = 2
+
+      50.times do
+        book = Work.new(title: "Title#{x}", category: "Book")
+        book.save
```
Comment (droberts-sea) [inline test/models/work_test.rb:114]: I like this idea - test it by making a bunch of books in a loop. Good work!

File: test/models/work_test.rb [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+      end
+    end
+
+  end
+
+
+  describe 'top_ten' do
+
+    it 'returns only 10 entries of each category' do
+
```
Comment (droberts-sea) [inline test/models/work_test.rb:109]: I would also be interested in the following cases:

- What if there are no works of that category?
- What if there's a tie for last place, e.g. works 9, 10 and 11 all have 0 votes?

File: test/models/work_test.rb [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+      ten_books = Work.top_ten("Book")
+
+      expect(ten_books.first.votes.count >= ten_books.last.votes.count).must_equal true
+    end
+  end
+
+  describe 'top_media' do
+
+    it 'has the most votes out of all media' do
+
```
Comment (droberts-sea) [inline test/models/work_test.rb:139]: I would also be interested in:

- What happens if there are no works?
- What happens if there are works but no votes?
- What happens if two works have the same number of votes?

**Comment by droberts-sea** [comment]
# Media Ranker
## What We're Looking For

Feature | Feedback
---     | ---
**Core Requirements** |
Git hygiene | yes
Comprehension questions	| yes
**General** |
Rails fundamentals (RESTful routing, use of named paths) | some extra routes
Views are well-organized (DRY, use of semantic HTML, use of partials) | semantic HTML is good, but views are not DRY - make sure to review partials
Errors are reported to the user | yes
Business logic lives in the models | yes
Models are thoroughly tested, including relations, validations and any custom logic | good start, missing some edge cases
**Wave 1 - Media** |
Splash page shows the three media categories | yes
Basic CRUD operations on media are present and functional | Can't delete a work with votes!
**Wave 2 - Users and Votes** |
Users can log in and log out | yes
The ID of the current user is stored in the session | yes
A user cannot vote for the same media more than once | no
All media lists are ordered by vote count | yes
Splash page contains a media spotlight | yes
**Wave 3 - Users and Votes** |
Media pages contain lists of voting users | yes
Individual user pages and the user list are present | yes
**Optional - Styling** |
Bootstrap is used appropriately | nope - make sure to get some practice with this on API muncher if you didn't on bEtsy
Look and feel is similar to the original | workflow is similar
**Overall** | Great job overall! Your implementation matches the demo site very closely, and I would say the learning goals for this assignment were definitely met. There are a few places where things could be cleaned up or your test coverage could be expanded, which I've tried to outline below, but in general I am quite happy with this submission. Keep up the hard work!
